cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,483 @@
+  {
+    total_feats     = num_feats;
+    avg_feat_length = avg_feat_len;
+    k               = k_param;
+    b               = b_param;
+  }
+
+  float __device__ operator()(const T2& value, const T2& num_feats_id_occ, const T2& feat_length)
+  {
+    T2 tf  = T2(value / feat_length);
+    T2 idf = raft::log<T2>(total_feats / num_feats_id_occ);
+    T2 bm  = ((k + 1) * tf) / (k * ((1.0f - b) + b * (feat_length / avg_feat_length)) + tf);
+
+    return idf * bm;
+  }
+  T2 avg_feat_length;
+  T1 total_feats;
+  T2 k;
+  T2 b;
+};
+
+/**
+ * @brief Calculates the tfidf values for a target matrix. Term frequency is calculate using
+ * logrithmically scaled frequency.
+ * @param total_feats_param: The total number of features in the matrix
+ */
+template <typename T1, typename T2>
+struct tfidf {
+  tfidf(T1 total_feats_param) { total_feats = total_feats_param; }
+
+  float __device__ operator()(const T2& value, const T2& num_feats_id_occ, const T2& feat_length)
+  {
+    T2 tf  = T2(value / feat_length);
+    T2 idf = raft::log<T2>(total_feats / num_feats_id_occ);
+    return tf * idf;
+  }
+  T1 total_feats;
+};
+
+template <typename T>
+struct mapper {
+  mapper(raft::device_vector_view<T> map) : map(map) {}
+
+  float __device__ operator()(const T& value)
+  {
+    T new_value = map[value];
+    if (new_value) {
+      return new_value;
+    } else {
+      return 0.0;
+    }
+  }
+
+  raft::device_vector_view<const T> map;
+};
+
+/**
+ * @brief Get unique counts
+ * @param handle: raft resource handle
+ * @param sort_vector: Input COO array that contains the keys.
+ * @param secondary_vector: Input with secondary keys of COO, (columns or rows).
+ * @param data: Input COO values array.
+ * @param itr_vals: Input array used to calculate counts.
+ * @param keys_out: Output array with one entry for each key. (same size as counts_out)
+ * @param counts_out: Output array with cumulative sum for each key. (same size as keys_out)
+ */
+template <typename T1, typename T2, typename IdxT>
+void get_uniques_counts(raft::resources& handle,
+                        raft::device_vector_view<T1, IdxT> sort_vector,
+                        raft::device_vector_view<T1, IdxT> secondary_vector,
+                        raft::device_vector_view<T2, IdxT> data,
+                        raft::device_vector_view<T2, IdxT> itr_vals,
+                        raft::device_vector_view<T1, IdxT> keys_out,
+                        raft::device_vector_view<T2, IdxT> counts_out)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  raft::sparse::op::coo_sort(sort_vector.size(),
+                             secondary_vector.size(),
+                             data.size(),
+                             sort_vector.data_handle(),
+                             secondary_vector.data_handle(),
+                             data.data_handle(),
+                             stream);
+
+  thrust::reduce_by_key(raft::resource::get_thrust_policy(handle),
```

Comment:
Ended up using the thrust version because it could handle vectors, which allows me to use the same code for both the csr and coo matrix versions of the encoding logic. Also the raft version does not support sparse matrix versions.

Reply:
Is this to compute the degree of each row in the sparse format? We have routines for this already. We have a `coo_degree` function [here](https://github.com/rapidsai/raft/blob/branch-24.10/cpp/include/raft/sparse/linalg/degree.cuh). Degree computation for CSR is actually really trivial- since you already have an array of offsets, you don't even need to count the columns because you can literally just diff the array (e.g. compute the difference between each value in the indptr array and the value that occurred before it).

Reply:
When we were using this function for rows, `coo_degree` was absolutely the right play. I was just trying to follow code reuse, but that ended up causing problems with larger datasets (in the form of illegal memory access errors). I have made it so this function is only used when we are trying to get a column-wise sum of the values (not just checking if there is a value like with rows). And we cant just use l1 normalization because I need the avg column size across all columns and the individual column avg. The `reduce by key` functions available in raft are for dense matrices only. This is why I have opted to use the  `reduce_by_key` when we are doing the column based processing.

Reply:
I think I may have not explained what is happening here correctly. `coo_degree` I think works as expected. I used the reduce_by_key because I am not just counting the number of occurrences for each column. I need to get the sum of values for each column. So that I know the total number of terms related to that document (for BoW problem set). And I cant just L1 normalize because I also need to calculate the avg number of terms in a document for across all documents as well as individual column average. We just dont have a substitute for `reduce_by_key` in raft, that handles sparse matrices.

Reply:
In that case, please create an issue to expose a RAFT function in the sparse (or dense) APIs that would remove the need to invoke thrust for this purpose, and reference it in a comment in the code. 

We need to follow up with this- we are trying very hard to keep our code clean and not introduce direct calls to thrust when it can be avoided. A segmented reduction for a coo formatted sparse matrix is a common operation and we should be encapsulating and centralizing that logic behind raft APIs. 

It’d be nice if we could expose that in this PR, but I don’t think it’s necessary. At the very least, we should creating issues so we can follow up when we have some time to pay down tech debt. 

This centralization serves another purpose also- when we find that we have functions used in several places, it provides a great well-bounded evaluation case for our devtechs to come in an optimize.

Reply:
I also think this operation is pretty straightforward to do in a sparse reduction routine- coo is a trivial case that can be done with simple atomics (and I’m pretty sure thrust does it this way also).

Reply:
Created https://github.com/rapidsai/raft/issues/2477. All sparse reduction routines in the raft API currently are row specific with CSR matrices https://docs.rapids.ai/api/raft/stable/cpp_api/sparse_linalg/#sparse-linear-algebra. I dont think adding the overhead of transforming the graph to CSC is the right move.